dual_net.py

```diff
@@ -162,21 +172,30 @@ def __call__(self, inp):
         Gets a feed-forward prediction for a batch of input boards of shape set
         during initialization.
         """
+        move_legality_mask = np.zeros(shape=(inp.shape[0], self.action_size))
+        for i in range(inp.shape[0]):
+            move_legality_mask[i] = self.env.get_legality_mask(inp[i])
         policy, value = self.sess.run([self.policy_predict, self.value_predict],
-                                      feed_dict={self.board_placeholder: inp})
+                                      feed_dict={self.board_placeholder: inp,
+                                                 self.move_legality_mask: move_legality_mask})
         return policy, value
 
-    def train(self, boards, pi, z):
+    def train(self, states, pi, z, legality_mask):
         """
         Performs one step of gradient descent based on a batch of input boards,
         MCTS policies, and rewards of shape [None, 1].  Shapes of inputs and policies
         should match input_shape and action_size as set during initialization.
         returns the batch loss
         """
-        self.sess.run([self.update_op], feed_dict={self.board_placeholder: boards,
+        move_legality_mask = np.zeros(shape=(len(states), self.action_size))
+        for i in range(len(states)):
+            move_legality_mask[i] = self.env.get_legality_mask(states[i])
+        self.sess.run([self.update_op], feed_dict={self.board_placeholder: states,
                                                    self.pi: pi,
-                                                   self.z: z})
-        loss = self.sess.run([self.loss], feed_dict={self.board_placeholder: boards,
+                                                   self.z: z,
+                                                   self.move_legality_mask: move_legality_mask})
+        loss = self.sess.run(self.loss, feed_dict={self.board_placeholder: states,
```

Comment:
We call sess.run twice -- once to update the weights and once to get the loss.  We can do this in 1 swing.